### PR TITLE
Label more fences

### PR DESCRIPTION
**Pull request by DavidPoliakoff**
This goes through a lot of Kokkos and just finds unlabeled fences and labels them

### CODE REVIEW

**Comment by DavidPoliakoff** [comment]
Note to reviewers: please do check my names here. I've done my best to spot obvious dumb stuff, but I might have missed things

File: containers/src/Kokkos_DualView.hpp [hunk 1]
```diff
@@ -597,8 +597,10 @@ class DualView : public ViewTraits<DataType, Arg1Type, Arg2Type, Arg3Type> {
     }
     if (std::is_same<typename t_host::memory_space,
                      typename t_dev::memory_space>::value) {
-      typename t_dev::execution_space().fence();
-      typename t_host::execution_space().fence();
+      typename t_dev::execution_space().fence(
+          "Kokkos::DualView<>::sync: fence after syncing DualView");
+      typename t_host::execution_space().fence(
+          "Kokkos::DualView<>::sync:fence after syncing DualView");
```
Comment (dalg24) [inline containers/src/Kokkos_DualView.hpp:603]:
```suggestion
          "Kokkos::DualView<>::sync: fence after syncing DualView");
```

File: containers/src/Kokkos_DynRankView.hpp [hunk 2]
```diff
@@ -1711,10 +1715,13 @@ inline void deep_copy(
 
   using src_traits       = ViewTraits<ST, SP...>;
   using src_memory_space = typename src_traits::memory_space;
-  Kokkos::fence();
+  Kokkos::fence(
+      "Kokkos::deep_copy(value_type, DynRankView): fence before copying "
+      "values");
```
Comment (dalg24) [inline containers/src/Kokkos_DynRankView.hpp:1720]: Technically there is only one value here ;)

File: containers/src/Kokkos_DynRankView.hpp [hunk 3]
```diff
@@ -1817,29 +1832,43 @@ inline void deep_copy(
...
+          "remapping views of incompatible shape");
     } else {
       Kokkos::Impl::throw_runtime_exception(
           "deep_copy given views that would require a temporary allocation");
     }
   } else {
-    Kokkos::fence();
+    Kokkos::fence(
+        "Kokkos::Impl::DeepCopy(DynRankView, DynRankView): fence due to same "
+        "src and dest");
```
Comment (dalg24) [inline containers/src/Kokkos_DynRankView.hpp:1871]: dest -> dst (?)

File: core/src/Kokkos_OpenMPTargetSpace.hpp [hunk 4]
```diff
@@ -248,7 +248,9 @@ struct DeepCopy<Kokkos::Experimental::OpenMPTargetSpace,
                                        omp_get_default_device()));
   }
   DeepCopy(const ExecutionSpace& exec, void* dst, const void* src, size_t n) {
-    exec.fence();
+    exec.fence(
+        "Kokkos::Impl::DeepCopy<OpenMPTarget, OpenMPTarget>: fence before "
```
Comment (dalg24) [inline core/src/Kokkos_OpenMPTargetSpace.hpp:252]: Should it be OpenTarget[Space]?

File: core/src/SYCL/Kokkos_SYCL_DeepCopy.hpp [hunk 5]
```diff
@@ -110,12 +110,16 @@ struct DeepCopy<Kokkos::Experimental::SYCLDeviceUSMSpace,
   }
 
   DeepCopy(const ExecutionSpace& exec, void* dst, const void* src, size_t n) {
-    exec.fence();
+    exec.fence(
+        "Kokkos::Impl::DeepCopy<SYCLDeviceUSMSpace, "
+        "SyclDeviceUSMSpace>::DeepCopy> fence before deep_copy");
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_DeepCopy.hpp:115]:
```suggestion
        "SYCLDeviceUSMSpace>: fence before deep_copy");
```

File: core/src/SYCL/Kokkos_SYCL_DeepCopy.hpp [hunk 6]
```diff
@@ -110,12 +110,16 @@ struct DeepCopy<Kokkos::Experimental::SYCLDeviceUSMSpace,
...
     DeepCopy<Kokkos::Experimental::SYCLDeviceUSMSpace,
              Kokkos::Experimental::SYCLDeviceUSMSpace,
              Kokkos::Experimental::SYCL>(Kokkos::Experimental::SYCL(), dst, src,
                                          n);
-    Kokkos::Experimental::SYCL().fence();
+    Kokkos::Experimental::SYCL().fence(
+        "Kokkos::Impl::DeepCopy<SYCLDeviceUSMSpace, "
+        "SyclDeviceUSMSpace>::DeepCopy: fence after deep_copy");
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_DeepCopy.hpp:122]:
```suggestion
        "SYCLDeviceUSMSpace>: fence after deep_copy");
```

File: containers/src/Kokkos_DynRankView.hpp (hunk 2 above)
Comment (DavidPoliakoff) [reply containers/src/Kokkos_DynRankView.hpp]: 402ea6a

File: containers/src/Kokkos_DynRankView.hpp (hunk 3 above)
Comment (DavidPoliakoff) [reply containers/src/Kokkos_DynRankView.hpp]: 402ea6a

File: core/src/Kokkos_OpenMPTargetSpace.hpp (hunk 4 above)
Comment (DavidPoliakoff) [reply core/src/Kokkos_OpenMPTargetSpace.hpp]: 402ea6a

**Review by dalg24 — APPROVED** [review]

**Review by Rombur — APPROVED** [review]

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 7]
```diff
@@ -420,20 +420,26 @@ class UnorderedMap {
   bool begin_erase() {
     bool result = !erasable();
     if (is_insertable_map && result) {
-      execution_space().fence();
+      execution_space().fence(
+          "Kokkos::UnorederedMap::begin_erase: fence before setting erasable "
+          "flag");
```
Comment (masterleinad) [inline containers/src/Kokkos_UnorderedMap.hpp:425]:
```suggestion
      execution_space().fence(
          "Kokkos::UnorderedMap::begin_erase: fence before setting erasable "
          "flag");
```

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 8]
```diff
@@ -420,20 +420,26 @@ class UnorderedMap {
...
       set_flag(erasable_idx);
-      execution_space().fence();
+      execution_space().fence(
+          "Kokkos::UnorederedMap::begin_erase: fence after setting erasable "
+          "flag");
```
Comment (masterleinad) [inline containers/src/Kokkos_UnorderedMap.hpp:429]:
```suggestion
      execution_space().fence(
          "Kokkos::UnorderedMap::begin_erase: fence after setting erasable "
          "flag");
```

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 9]
```diff
@@ -420,20 +420,26 @@ class UnorderedMap {
...
     }
     return result;
   }
 
   bool end_erase() {
     bool result = erasable();
     if (is_insertable_map && result) {
-      execution_space().fence();
+      execution_space().fence(
+          "Kokkos::UnorederedMap::end_erase: fence before erasing");
```
Comment (masterleinad) [inline containers/src/Kokkos_UnorderedMap.hpp:438]:
```suggestion
      execution_space().fence(
          "Kokkos::UnorderedMap::end_erase: fence before erasing");
```

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 10]
```diff
@@ -420,20 +420,26 @@ class UnorderedMap {
...
       Impl::UnorderedMapErase<declared_map_type> f(*this);
       f.apply();
-      execution_space().fence();
+      execution_space().fence(
+          "Kokkos::UnorederedMap::end_erase: fence after erasing");
```
Comment (masterleinad) [inline containers/src/Kokkos_UnorderedMap.hpp:442]:
```suggestion
      execution_space().fence(
          "Kokkos::UnorderedMap::end_erase: fence after erasing");
```

File: core/src/Kokkos_OpenMPTargetSpace.hpp [hunk 11]
```diff
@@ -248,7 +248,10 @@ struct DeepCopy<Kokkos::Experimental::OpenMPTargetSpace,
                                        omp_get_default_device()));
   }
   DeepCopy(const ExecutionSpace& exec, void* dst, const void* src, size_t n) {
-    exec.fence();
+    exec.fence(
+        "Kokkos::Impl::DeepCopy<OpenMPTargetSpace, OpenMPTargetSpace>: fence "
+        "before "
+        "copy");
```
Comment (masterleinad) [inline core/src/Kokkos_OpenMPTargetSpace.hpp:254]:
```suggestion
        "Kokkos::Impl::DeepCopy<OpenMPTargetSpace, OpenMPTargetSpace>: fence before copy");
```
I'm sure `clang-format` can do better.

**Review by masterleinad — CHANGES_REQUESTED** [review]

File: core/src/Kokkos_OpenMPTargetSpace.hpp (hunk 11 above)
Comment (DavidPoliakoff) [reply core/src/Kokkos_OpenMPTargetSpace.hpp:254]: This is what clang-format leads to. I'm not going to do clang-format _and then_ manually polish the results. Other comments look good though

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 10 above)
Comment (DavidPoliakoff) [reply containers/src/Kokkos_UnorderedMap.hpp]: fa62ab9

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 9 above)
Comment (DavidPoliakoff) [reply containers/src/Kokkos_UnorderedMap.hpp]: fa62ab9

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 8 above)
Comment (DavidPoliakoff) [reply containers/src/Kokkos_UnorderedMap.hpp]: fa62ab9

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 7 above)
Comment (DavidPoliakoff) [reply containers/src/Kokkos_UnorderedMap.hpp]: fa62ab9

File: core/src/Kokkos_OpenMPTargetSpace.hpp (hunk 11 above)
Comment (masterleinad) [reply core/src/Kokkos_OpenMPTargetSpace.hpp:254]: Just as I expected, I get
```suggestion
        "Kokkos::Impl::DeepCopy<OpenMPTargetSpace, OpenMPTargetSpace>: fence "
        "before copy");
```
with `clang-format` after putting everything on one line. I don't care enough to not approve, though.

**Review by masterleinad — APPROVED** [review]

**Comment by DavidPoliakoff** [comment]
The EventCorrectness test is actually my fault, I didn't test an HPX build. Just change the default instance ID to 1

```c++
  uint32_t m_instance_id = 0; // should be 1
```

And all the other places m_instance_id gets set to `0`, and you'll be in line with other implementations. Apologies

**Pull request merged**